### PR TITLE
[MilkmApp] Handle states into history table

### PR DESCRIPTION
**Pull request by yvesfracari**
https://www.loom.com/share/495730bdbcc44b5782e1c6dbb27c0103

### CODE REVIEW

File: apps/cow-tools/codegen.ts [hunk 1]
```diff
@@ -3,9 +3,9 @@ import type { CodegenConfig } from "@graphql-codegen/cli";
 const config: CodegenConfig = {
   overwrite: true,
   schema: "https://milkman-api.up.railway.app/",
-  documents: "src/**/*.tsx",
+  documents: "src/**/*.ts",
   generates: {
-    "src/gql/generated.ts": {
+    "src/lib/gql/generated.ts": {
```
Comment (luizakp) [inline apps/cow-tools/codegen.ts:8]: why is this needed?

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  const publicClient = createPublicClient({
+    chain: goerli,
+    transport: http(),
+  });
+
+  useEffect(() => {
+    async function loadOrders() {
+      const { users } = await sdk.AllTransactionFromUser({
+        user: safe.safeAddress,
+      });
```
Comment (luizakp) [inline apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:87]: this is getting all transactions from safe, not only from milkman, right?

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+      const hasTokenBySwap = tokenBalances.map(
+        (tokenBalance) => tokenBalance > 0,
+      );
+
+      const cowOrdersBySwap = await Promise.all(
+        orderContractsBySwap.map((orderContract) =>
+          getCowOrders(orderContract as Address),
+        ),
+      );
```
Comment (luizakp) [inline apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:130]: here is where we check if the transaction is in fact from milkman?

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            id
+            name
+            symbol
+            decimals
+          }
+        }
+      }
+    }
+  }
+`);
```
Comment (luizakp) [inline apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:46]: think this not needed anymore
```suggestion

```

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
+import { Address } from "@bleu-fi/utils";
+import { useSafeAppsSDK } from "@gnosis.pm/safe-apps-react-sdk";
+import { gql } from "graphql-tag";
+import { useEffect, useState } from "react";
+import { createPublicClient, http, PublicClient } from "viem";
+import { goerli } from "viem/chains";
+
+import { erc20Abi } from "#/lib/abis/erc20";
```
Comment (luizakp) [inline apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:8]: import { erc20ABI } from "@wagmi/core"; didn't work?

File: apps/cow-tools/src/lib/abis/erc20.ts
Comment (luizakp) [inline apps/cow-tools/src/lib/abis/erc20.ts:1]: if `import { erc20ABI } from "@wagmi/core"` works we can delete this

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 6]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
+  return (
+    <>
+      <Table.BodyRow key={transaction.id}>
+        <Table.BodyCell>
+          <button onClick={() => setShowOrdersRows(!showOrdersRows)}>
+            {showOrdersRows ? (
+              <ArrowUpIcon className="h-5 w-5 text-blue9 hover:text-blue10" />
+            ) : (
+              <ArrowDownIcon className="h-5 w-5 text-blue9 hover:text-blue10" />
+            )}
```
Comment (luizakp) [inline apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:185]:
```suggestion
            {showOrdersRows ? (
              <ChevronUpIcon className="h-5 w-5 text-blue9 hover:text-blue10" />
            ) : (
              <ChevronDownIcon className="h-5 w-5 text-blue9 hover:text-blue10" />
            )}
```

image

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 7]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
-          <Table.BodyCell>Test 2</Table.BodyCell>
-          <Table.BodyCell>
-            {TransactionStatus.CANCELATION_TO_BE_EXECUTED}
-          </Table.BodyCell>
-          <Table.BodyCell>
-            <CancelButton
-              status={TransactionStatus.CANCELATION_TO_BE_EXECUTED}
-            />
-          </Table.BodyCell>
-        </Table.BodyRow>
```
Comment (luizakp) [inline apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:71]: thanks 😅

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 8]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
+            )}
+          </div>
+        </Table.BodyCell>
+        <Table.BodyCell>
+          <CancelButton
+            disabled={[
+              TransactionStatus.EXECUTED,
+              TransactionStatus.CANCELED,
+              TransactionStatus.EXECUTED_AND_CANCELED,
+            ].includes(transactionStatus)}
```
Comment (luizakp) [inline apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:231]: 🔝

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 9]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
+  }
+
+  const swapStatus = transaction.swaps.map((swap, i) =>
+    getSwapStatus(hasToken?.[i], cowOrders?.[i]),
+  );
+  const transactionStatus = getTransactionStatus(swapStatus);
   const txUrl = buildBlockExplorerTxUrl({
-    chainId: order.chainId,
-    txHash: order.transactionHash,
+    chainId: transaction.swaps[0].chainId,
```
Comment (luizakp) [inline apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:154]: this is because there is no way we can have a multichain view, right? maybe leave a comment for the future

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 10]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
+          />
+        </Table.BodyCell>
+      </Table.BodyRow>
+      {showOrdersRows &&
+        transaction.swaps.map((order, index) => (
+          <TableRowOrder
+            key={order.id}
+            order={order}
+            orderStatus={swapStatus[index]}
+          />
```
Comment (luizakp) [inline apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:241]: not a fan of using index 🤔

**Review by luizakp — DISMISSED** [review]
LGTM, just left some questions and suggestions

File: apps/cow-tools/codegen.ts (hunk 1 above)
Comment (yvesfracari) [reply apps/cow-tools/codegen.ts:8]: one query was on .ts file, that wasn't watched by the codegen. Also, the path for the generated file was wrong. Regarding the .ts and .tsx I think that I can add both here.

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts (hunk 2 above)
Comment (yvesfracari) [reply apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:87]: Actually, the `sdk` is our ponder api. So it is getting, just the milkman.

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts (hunk 3 above)
Comment (yvesfracari) [reply apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:130]: Actually, here we check if one order was posted on cow. I can add some comments on these parts.

File: apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts (hunk 4 above)
Comment (yvesfracari) [reply apps/cow-tools/src/hooks/useUserMilkmanTransactions.ts:46]: Actually, it is needed 😶‍🌫️  to get all

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx [hunk 11]
```diff
@@ -53,38 +60,211 @@ export function OrderTable({ orders }: { orders: AllSwapsQuery["swaps"] }) {
...
+    if (allOrdersPlaced) {
+      return TransactionStatus.ORDER_PLACED;
+    }
+
+    return TransactionStatus.MILKMAN_CREATED;
+  }
+
+  const swapStatus = transaction.swaps.map((swap, i) =>
+    getSwapStatus(hasToken?.[i], cowOrders?.[i]),
+  );
```
[the remark this answers is not in the document]
Comment (yvesfracari) [reply apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:151]: We can create a custom object that has it all. It will be easier to use.

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx (hunk 9 above)
Comment (yvesfracari) [reply apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:154]: The app is connected to just your wallet (one chain), so I think that a multichain view doesn't make sense in this case. But let me know your suggestions

File: apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx (hunk 10 above)
Comment (yvesfracari) [reply apps/cow-tools/src/app/milkman/(components)/OrdersTable.tsx:241]: yeah, I get your point, this can lead to some errors. I'll try to create one object that has all the data that we need.
